ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkId.java

```diff
@@ -0,0 +1,378 @@
+import org.hyperledger.besu.ethereum.rlp.RLPInput;
+import org.hyperledger.besu.ethereum.rlp.RLPOutput;
+import org.hyperledger.besu.util.bytes.Bytes32;
+import org.hyperledger.besu.util.bytes.BytesValue;
+import org.hyperledger.besu.util.bytes.BytesValues;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.zip.CRC32;
+
+public class ForkId {
+
+  private Hash genesisHash;
+  private CRC32 crc = new CRC32();
+  private Long currentHead;
+  private Long forkNext;
+  private Long highestKnownFork = 0L;
+  private ForkIdEntry lastKnownEntry;
+  private boolean useForkId;
+  private ArrayDeque<ForkIdEntry> forkAndHashList;
+
+  public ForkId(
+      final Hash genesisHash,
+      final Set<Long> forks,
+      final Long currentHead,
+      final boolean useForkId) {
+    this.genesisHash = genesisHash;
+    this.currentHead = currentHead;
+    this.useForkId = useForkId;
+    if (forks != null) {
+      forkAndHashList = collectForksAndHashes(forks, currentHead);
+    } else {
+      forkAndHashList = new ArrayDeque<>();
+    }
+  };
+
+  public static ForkId buildCollection(
+      final Hash genesisHash,
+      final List<Long> forks,
+      final Blockchain blockchain,
+      final List<Capability> caps) {
+    if (forks == null) {
+      return new ForkId(genesisHash, null, blockchain.getChainHeadBlockNumber(), false);
+    } else {
+      Set<Long> forkSet = new LinkedHashSet<>(forks);
+      return new ForkId(
+          genesisHash,
+          forkSet,
+          blockchain.getChainHeadBlockNumber(),
+          caps.contains(Capability.create(EthProtocol.NAME, EthProtocol.EthVersion.V64)));
+    }
+  };
+
+  public static ForkId buildCollection(final Hash genesisHash, final List<Long> forks) {
+    if (forks == null) {
+      return new ForkId(genesisHash, null, Long.MAX_VALUE, false);
+    } else {
+      Set<Long> forkSet = new LinkedHashSet<>(forks);
+      return new ForkId(genesisHash, forkSet, Long.MAX_VALUE, true);
+    }
+  };
+
+  public static ForkId buildCollection(final Hash genesisHash) {
+    return new ForkId(genesisHash, null, Long.MAX_VALUE, false);
+  };
+
+  public static ForkIdEntry readFrom(final RLPInput in) {
+    in.enterList();
+    final String hash = in.readBytesValue(BytesValues::asString);
+    final long next = in.readLong();
+    in.leaveList();
+    return new ForkIdEntry(hash, next);
+  }
+
+  // Non-RLP entry (for tests)
+  public static ForkIdEntry createIdEntry(final String hash, final long next) {
+    return new ForkIdEntry(hash, next);
+  }
+
+  public ArrayDeque<ForkIdEntry> getForkAndHashList() {
+    return this.forkAndHashList;
+  }
+
+  public Hash getLatestForkId(final int protocolVersion) {
+    // TODO: implement handling for forkID in status message
+    if (useForkId && protocolVersion > 63) {
```

Comment:
I think this logic should go up a level, to `handleNewConnection`, if possible.

Reply:
Oh, true.  That makes more sense.